### ISSUE TITLE
Add generalized wait_event() method

### ISSUE THREAD

**Issue by lhecker** [zonyitoo/coio-rs#22]
Currently the `Scheduler::wait_event()` method is a hidden one and only used internally.
I propose to simplify the `wait_event()` method to make it possible for the users of coio to await mio events in a coroutine.

This feature should be added as generalized as possible though. It should be possible, for instance, to compose `wait_event()` with other asynchronous events of any kind.

For instance this feature could be extended to feature a `wait_event_until(..., timeout)` method, which would make it possible to specify the maximum duration to wait for that event, by fusing `wait_event()` together with a timer. If one would like to write a HTTP client/server for instance this feature would make it easily possible to read from the network connection while keeping the HTTP timeout in mind (for instance a server is usually keeping connections around for a certain time to implement the Keep-Alive feature).

**Comment by jedisct1** [comment]
Is there currently a way to enforce a timeout on `TcpStream` reads or is it also something blocked by this issue?

**Comment by zonyitoo** [comment]
I think `TcpStream` already has a method to set timeout. Please refer to [set_keepalive](http://rustdoc.s3-website-us-east-1.amazonaws.com/mio/v0.5.x/mio/tcp/struct.TcpStream.html#method.set_keepalive) method in Mio's `TcpStream`. Coio's `TcpStream` is just a wrapper on Mio's.

**Comment by jedisct1** [comment]
Already tried it, but it doesn't do anything, at least under coio :/

**Comment by zonyitoo** [comment]
It should work .. I have tried it in my another project. It's argument is in seconds, not milliseconds.

**Comment by lhecker** [comment]
Guys you got something wrong there... TCP Keep-Alive is for detecting broken connections. That is: Connections where the final FIN packet didn't reach the other peer - for instance if you pull your network cable). This makes it possible to let read() return immediately instead of waiting forever for some data from the other peer (which will never come).

What @jedisct1 needs (if I understand you correctly) is some functionality to actually kill old connections to prevent resource exhaustion and thus DoS. And this is only possible (in a deterministic way) if we're able to terminate any calls to `read()` (which may block indefinitely if the peer is never sending anything).
For `read()`s the current coroutine is parked and stored in the EventLoop until a read event is received. It's thus necessary to rewrite the current code to allow for coroutines to be taken from the EventLoop before the read (or whatever) event is fired. This problem is basically the same this I'd like to achieve in this issue anyway (=> read/write timeouts).

Apart from the usual complications of "How do I do this without using Mutex<> everywhere?" my biggest problem is:

Is it okay to create a small-scale solution, which only works in some cases? It would be cool to have a generic solution instead, which would allow us to freely compose different kinds of "promises" together. For instance if I wan't to solve @jedisct1's problem:
```rust
Scheduler::spawn(|| {
    loop {
        let mut buf = [0u8; 1024];
        let result = await_one!(kill_signal.await(), client.read(&mut buf));
        match result {
            Result0(..) => return,
            Result1(len) => println!("{}", &buf[..len]),
        }
    }
});
```
(`await_one!()` would be a macro which returns as soon as one of the N passed promises is fulfilled and cancel the remaining one)
This could be applied to _everything_ else and thus provide incredible flexibility as soon as we would like to compose different things together. For instance we could also `recv()` from 2 or more different channels at once and thus create a loop which composes N channels to 1.
On the other hand this change would be quite viral... (because async. methods would return promises instead of being "suspend-down" as they currently are) :disappointed:

So yeah... I'd be glad if you guys (or rather: you @zonyitoo :smile:) got any idea how to flexibly compose different asynchronous operations together...

**Comment by zonyitoo** [comment]
I have an suggestion taken from the source code in one of our private libraries. The basic idea is

1. When the socket is going to wait for event (such as calling the `read` syscall), then first register it to the event loop, after that,
2. register a timeout for this coroutine

So if the `read` event was comeback as it was expected, then the coroutine is waken up to process this event, otherwise, the coroutine is waken up by the timer and got a timeout error.

This should be one of the proper solutions that meets @jedisct1 's requirement.
